pkg/types/querybuildertypes/querybuildertypesv5/json_decoder.go

```diff
@@ -0,0 +1,163 @@
+package querybuildertypesv5
+
+import (
+	"bytes"
+	"encoding/json"
+	"reflect"
+	"strings"
+
+	"github.com/SigNoz/signoz/pkg/errors"
+	"github.com/SigNoz/signoz/pkg/types/telemetrytypes"
+)
+
+// UnmarshalJSONWithSuggestions unmarshals JSON data into the target struct
+// and provides field name suggestions for unknown fields
+func UnmarshalJSONWithSuggestions(data []byte, target any) error {
+	return UnmarshalJSONWithContext(data, target, "")
+}
+
+// UnmarshalJSONWithContext unmarshals JSON with context information for better error messages
+func UnmarshalJSONWithContext(data []byte, target any, context string) error {
+	// First, try to unmarshal with DisallowUnknownFields to catch unknown fields
+	dec := json.NewDecoder(bytes.NewReader(data))
+	dec.DisallowUnknownFields()
+
+	err := dec.Decode(target)
+	if err == nil {
+		// No error, successful unmarshal
+		return nil
+	}
+
+	// Check if it's an unknown field error
+	if strings.Contains(err.Error(), "unknown field") {
```

Comment:
```suggestion
	if strings.Contains(err.Error(), "unknown field") {
		return err
	}
```
This usecase is ripe for defining a custom error code.

You can define a ErrCodeQuerierUnknownField = errors.MustNewCode("querier_unknown_field") and then call errors.Asc().

It's not very maintainable when you call string.Contains.

Reply:
I wouldn't have used `string.Contains` if I had an option. This is an error that comes from the JSON library. There exists an error sentinel for this in experimental v2 json, but not in current json from stdlib, so I had to go with `strings.Contains(err.Error(), "unknown field")`

Reply:
https://github.com/golang/go/blob/master/src/encoding/json/v2/errors.go#L39


https://github.com/golang/go/blob/master/src/encoding/json/decode.go#L737